File: rust/src/bin/step0_repl.rs
use mal::readline;

fn read(string: &str) -> &str {
    string
}

fn eval<'a>(ast: &'a str, _env: &str) -> &'a str {
    ast
}

fn print(exp: &str) -> &str {
    exp
}

fn rep(string: &str) -> &str {
    print(eval(read(string), ""))
}

fn main() {
    let mut prompt = readline::Prompt::new();

    while let Some(line) = prompt.readline("user> ") {
        let result = rep(&line);
        println!("{}", result);
    }
}

File: impls/noc7c9-typescript/src/step0_repl.ts
import * as readline from './readline';

function read(line: string): string {
    return line;
}

function eval_(line: string): string {
    return line;
}

function print(line: string): string {
    return line;
}

(function main() {
    const prompt = readline.initialize('user> ');
    let line;
    while ((line = prompt()) != null) {
        line = read(line);
        line = eval_(line);
        line = print(line);
        console.log(line);
    }

    process.exit(0);
})();
